Issues/374

noahl opened the pull request:
Introduce a function `new_fact` which defines a new fact and
automatically adds it to the list of all facts. This makes it
impossible to accidentally leave it out, preventing bug #373 from
happening again.

Also make `new_fact` function be the source of fact documentation, and
require documentation for every new fact, which will prevent bugs like

Finally, also make fact groups be inclusive rather than exclusive, to
match customer expectations.

Closes #374 . Closes #367 .

Commit:
Make fact definition more robust

Introduce a function `new_fact` which defines a new fact and
automatically adds it to the list of all facts. This makes it
impossible to accidentally leave it out, preventing bug #373 from
happening again.

Also make `new_fact` function be the source of fact documentation, and
require documentation for every new fact, which will prevent bugs like

Finally, also make fact groups be inclusive rather than exclusive, to
match customer expectations.

Commit:
Generate fact docs automatically

Commit:
Fix lint error

Also remove dead code from Makefile.

chambridge commented on rho/facts.py:
```diff
@@ -0,0 +1,242 @@
+    facts_to_collect.update(ALWAYS_COLLECT)
+
+    if not facts_to_collect.issubset(ALL_FACTS):
+        log.error('Bad fact list %s', facts_to_collect)
```
It would be nicer here if we output the difference that shows the facts that are actually invalid in the subset.

noahl replied:
Done, thanks! Sample output:
```
> PYTHONPATH=rho:$PYTHONPATH bin/rho scan --profile=jboss-test-server --facts jboss.eap.installed-versions --facts jboss.eap.deploy-dates --reportfile=out.csv -vvv --facts fizzbuzz --facts abc
Unknown facts requested: abc, fizzbuzz
```

chambridge reviewed (commented):
Looks good.

Need to update the Contributing.rst with these instructions:
https://github.com/quipucords/rho/blob/master/CONTRIBUTING.rst#adding-facts

elyezer reviewed (approved):
I really like these changes as it deduplicate a lot of work needed to get fact information across different rho subcommands and in the documentation.

Please wait for at least one more reviewer as this is a change that touches many files.

I will run Camayoc tests with this patch to make sure it continue returning the expected fact list.

Commit:
Update for review comments

Update CONTRIBUTING.rst with the new way to add facts, and print out
the list of bad facts in case of bad input. Thanks to chambrid for the
review.

noahl commented:
@chambridge I just uploaded a new version that updates CONTRIBUTING.rst and prints the list of bad facts when we have it. Please take a look.
